State / CPT: Create notices middleware for observing actions

aduth opened the pull request:
Related: #6536
Related: #6306

This pull request seeks to explore a new middleware for notices to enable monitoring action types and dispatching notice creation in response to those actions. The advantage of this approach over earlier iterations (#6306) is that in using the reducer for observing actions, we don't have access to the entire global state. This makes it difficult to customize the notice, for example including the title of a post when a delete fails, since the action only contains the post and site IDs.

Since the middleware signature allows access to the entire global state, we can make use of any existing selector to retrieve data outside of the action or notices state. It also simplifies and removes redundancy in creating the notices by reusing existing action creators and the reducer behaviors corresponding to those action types.

![Delete failure with title](https://cloud.githubusercontent.com/assets/1779930/16699579/e90cc446-4522-11e6-9777-a7ec66129da1.png)

__Testing instructions:__

Repeat testing instructions from #6536 

__Follow-up tasks:__

I still think it'd be worth exploring expanding this middleware to capture `meta` directly from the action, as mentioned at https://github.com/Automattic/wp-calypso/pull/6306#discussion_r69541313 .

/cc @mtias @gwwar @artpi 

Test live: https://calypso.live/?branch=add/notices-middleware

gwwar commented on client/state/notices/middleware.js:
```diff
@@ -0,0 +1,39 @@
+import { getSitePost } from 'state/posts/selectors';
+
+const OBSERVERS = {
+	[ POST_DELETE_FAILURE ]: [
```
What happens when we wish to have different notices that are keyed to the same actions? For example a failed delete in a post list, versus a failed delete in the Calypso editor?

aduth replied:
>For example a failed delete in a post list, versus a failed delete in the Calypso editor?

Not sure this is something we'd want to do in practice, though it should be do-able; you could, for example, switch on `state.ui.section`.

gwwar replied:
It'd be uncommon, but it's nice to have the flexibility to do so.

dmsnell replied:
This is the question I raised in #6262 - it seems like the burden of proof should like in asking, "when would we have a legitimate reason to have different notices for the same type of event."

gwwar commented:
The code here looks good. 👍 Having access to the full state tree is a good improvement. 

> I still think it'd be worth exploring expanding this middleware to capture meta directly from the action

I'm curious to see what this looks like. If we run with this middleware file, it'll grow in size pretty quickly and the notice logic will remain apart from its related sub-tree.

aduth commented:
>I'm curious to see what this looks like.

In my mind, one could include a `meta.notice` property on any action which mirrors the structure of the `notice` object created in `createNotice`:

https://github.com/Automattic/wp-calypso/blob/eb4252e/client/state/notices/actions.js#L22-L30

I don't know when we'd ever want to use it in this way, so likely best to wait until the need arises.

>If we run with this middleware file, it'll grow in size pretty quickly and the notice logic will remain apart from its related sub-tree.

I might argue this is a good thing. Same as in #6306, I quite like that there's a single file where one can observe the different action types which can trigger a notice. Tying to the original action creator also limits us to the same ceveats around global state. I suppose technically one could claim there's `getState` argument on the thunk signature, but there's a general consensus that this should be avoided.

I considered a "one-per-file" action type to notices behaviors, but quickly ran into the issue that the value of the action type could technically be different than its type name, making it difficult to name files statically. e.g. What if we were to start namespacing the values of our action types with a `wpcalypso/` prefix if/when we introduce the idea of plugins?

aduth commented:
Pushed a rebase which simplified observers by removing array declaration (premature). Also added tests and moved existing `notices/reducer.js` handlers to the new observer object. In the process ditched "count"*, but I think we'll not want this anyways, since we should have granular control from the notice to e.g. undo a trash action for an individual post. This will be implemented in a subsequent pull request.

_*_ Not that it's difficult to implement. In fact, here's a refactor of `createNotice` to allow arbitrary options to be stored in a notice:

```diff
diff --git a/client/state/notices/actions.js b/client/state/notices/actions.js
index b75aeb1..c0d4810 100644
--- a/client/state/notices/actions.js
+++ b/client/state/notices/actions.js
@@ -19,19 +19,22 @@ export function removeNotice( noticeId ) {
 }
 
 export function createNotice( status, text, options = {} ) {
-       const notice = {
-               noticeId: options.id || uniqueId(),
-               duration: options.duration,
-               showDismiss: ( typeof options.showDismiss === 'boolean' ? options.showDismiss : true ),
-               isPersistent: options.isPersistent || false,
-               displayOnNextPage: options.displayOnNextPage || false,
-               status: status,
-               text: text
-       };
+       const {
+               id: noticeId = uniqueId(),
+               showDismiss = true,
+               isPersistent = false,
+               displayOnNextPage = false
+       } = options;
 
        return {
                type: NOTICE_CREATE,
-               notice: notice
+               notice: {
+                       ...options,
+                       noticeId,
+                       showDismiss,
+                       isPersistent,
+                       displayOnNextPage
+               }
        };
 }
```

blowery commented on client/state/notices/middleware.js:
```diff
@@ -0,0 +1,62 @@
+} from 'state/action-types';
+
+export const OBSERVERS = {
+	[ POST_DELETE_FAILURE ]: ( action, dispatch, getState ) => {
```
so this file will grow to know about a bunch of different action types? 

maybe expose a way to register an observer from the outside? that way action creators and their notices could be bundled together...

blowery replied:
ah, now i see the existing discussion in the base ticket..

aduth replied:
>that way action creators and their notices could be bundled together...

Do you think there's more benefit it keeping them together this way? Or, more generally, do you have concerns about this file growing in size?

blowery replied:
a bit. the stuff that's in there now has no bearing on the reader, but it's being included for the reader bundle... it's not that large yet, but if we require all notices to be in here, it'll grow and grow..

gwwar replied:
javascript decorators are still stage-1, but if we had them available, I could see us keeping this logic closer together and annotating the class method so that it registers with this middleware

Maybe something very roughly like:
```javascript
@createNotice( POST_SAVE_SUCCESS )
postDeleteSuccess( action, dispatch ) {
    dispatch( successNotice( translate( 'Post successfully deleted' ) ) );
}
```

aduth replied:
>the stuff that's in there now has no bearing on the reader, but it's being included for the reader bundle

I wonder if this could be an argument against keeping them alongside the action creators. It seems sensible to me that the notices behavior should always be globally aware of action types it's concerned with, not dependent on the action creators file having been loaded beforehand (though presumably in most/all cases the action creator to have triggered the action will most likely have been loaded beforehand anyways).

>javascript decorators are still stage-1

I've a love/hate relationship with decorators. They seem like a neat idea, but different enough that I worry it's the type of proposal that won't ever be adopted, and would be painful to rip out of code if we were to start depending on it now. I'd be more comfortable waiting until it at least reaches stage 2.

gwwar replied:
> I'd be more comfortable waiting until it at least reaches stage 2.

Oh definitely, this was just an example of how we could arrange the code in the future if this ever stabalizes

artpi replied:
I have no concern with this growing in size whatsoever.
What concerns me here (and it's the same discussion I've been having with @dmsnell over the connection middleware) is the separation of concerns.
In my opinion, `middleware.js` should not be concerned with translations/logic specific to some actions.
I do not like this centralized mega-case statements or huge arrays.
For the same reason, I do not like "stylesheet" and "javascript' folder separation in some projects, because I think all component logic should be close to that component. When I'm adding / editing component, I should not be venturing outside that folder too often.

Granted, we did not settle upon correct idea to link these "middleware plugins" with main middleware - as I said, I try to come up with something in the connection middleware efforts.

dmsnell replied:
> I wonder if this could be an argument against keeping them alongside the action creators

as @artpi mentioned, this is the main difference between our API middleware PRs. I do not think that the reader should care or know anything about notices. why should it? the reader knows that it needs to tell Calypso to delete a post or something like that - end of story. it's now up to Calypso to run that action and respond accordingly. if it should fail, then that failure can and will trigger a notice. what if we're offline? do we need to make the reader know about online/offline status?

if the logic for actually accomplishing the action is spread out we will end up with hard to manage spaghetti, but if we split it out then we can take advantage of code reuse, simplification of the independent functional blocks, and management of global/central policies.

dmsnell replied:
> javascript decorators are still stage-1

we can already accomplish this goal with some code transformation, something @aduth I think you and I discussed with someone else (or more people too). the cool thing about using these macros (however you want to call them, come they from Babel or sweet.js or anywhere else) is that by keeping them in-syntax, we can make a system that works whether or not the macro is active

```js
const handlePostDeleteSuccess = ( … ) => …
createActionHandler( POST_DELETE_SUCCESS, handlePostDeleteSuccess );
```

we can write `createActionHandler` first as a simple JS function, then when we get fed-up or experimental, we can write a macro to transform that at compile-time to whatever the decorator would do.

I still vote to keep these handlers all centralized and in one place, even with the decorators.

artpi commented:
> In my mind, one could include a meta.notice property on any action which mirrors the structure of the notice object created in createNotice:

I like this a lot!

dmsnell commented on client/state/notices/middleware.js:
```diff
@@ -0,0 +1,62 @@
+};
+
+export default function noticesMiddleware( { dispatch, getState } ) {
+	return ( next ) => ( action ) => {
```
I'm guessing you prefer to not do this, but many of the Redux libraries tend to use fat-arrow functions and without parenthesis for single argument functions (not sure if this is experience or my personal bias giving me this impression). to me it looks far better than the verbose mixture of `function( … ) { return ( next ) => ( action ) => … }`

```js
export const noticesMiddleware = ( { dispatch, getState } ) => next => action => {
	…
}
```

also, this threw me for a loop, because I see different definitions for the middleware API depending on where I look. turns out this has been a controversial thing and although I found several PRs/issues dealing with changing the signature of `applyMiddleware`, I never found the commits that actually changed it `¯\_(ツ)_/¯`

aduth replied:
>many of the Redux libraries tend to use fat-arrow functions and without parenthesis for single argument functions

I'm not a huge fan of the single argument exception for just that reason - it's an exception. One which must be learned, and I don't find the benefit to outweigh this cost and the cost of losing a common pattern for argument declarations (i.e. delineated by parentheses). I also don't see it being too much an issue to mix the two styles of functions if it helps bring a good balance of readability and compactness.

I'd agree though that it is a documented style of Redux middleware to take the inlined arrow function approach, and so to meet those expectations it would make more sense to rewrite this similar to how you'd suggested.

I do have strong opinions on the readability of arrow functions generally, but we can save that debate for another day 😄

dmsnell replied:
Thanks for matching the redux style.

It's funny; I feel like the parentheses for multiple arguments are the exception due to JS syntax not dealing with this well...

```
const combine = a, b => [ a, b ]
```

We should have some kind of challenge at the GM: Halo tournament, survivalist race, eating contest, or some other kind of death match to resolve these types of matters :wink:

aduth replied:
>Halo tournament

Ooh, it's on!

dmsnell commented on client/state/notices/test/middleware.js:
```diff
@@ -0,0 +1,177 @@
+				notice: {
+					text: 'Hello World'
+				}
+			} );
```
I haven't seen this being done. It would be great to have a write-up on testing dispatched actions like this in the Calypso digest

cc: @mtias

mtias replied:
Good call.

dmsnell commented on client/state/notices/middleware.js:
```diff
@@ -0,0 +1,62 @@
+export default function noticesMiddleware( { dispatch, getState } ) {
+	return ( next ) => ( action ) => {
+		if ( OBSERVERS.hasOwnProperty( action.type ) ) {
+			OBSERVERS[ action.type ]( action, dispatch, getState );
```
do we have a good reason to switch around the function signature here? I guess it makes sense not to pass in `next()`, but it feels a bit weird going from `( { dispatch, getState } ) => next => action` to `( action, dispatch, getState )`

aduth replied:
>do we have a good reason to switch around the function signature here?

I was thinking of it more in terms of likelihood that the observer would want to use a particular argument. For example, two of the included observers don't need `getState` at all, so they can simplify their signature to `( action, dispatch )`. I guess I must have kept some of the consistency, because it's also the case that one might not need the action object either. I thought also of passing these an object with the shorthand, which allows the handler to opt-in to those "arguments" it cares for:

i.e.:

```js
OBSERVERS[ action.type ]( { action, dispatch, getState } );

// ...

[ POST_DELETE_SUCCESS ]: ( { dispatch } ) => {
	dispatch( successNotice( translate( 'Post successfully deleted' ) ) );
}
```

dmsnell replied:
> I thought also of passing these an object with the shorthand, which allows the handler to opt-in to those "arguments" it cares for:

I like what you did more than this. As much as I like this suntax for optional parameters, I feel like we risk trying to reimplement argument passing when we rely on it exclusively

aduth replied:
>I like what you did more than this.

Sorry, it's not clear, do you prefer the original approach from 2ae4aa5455f3079ee7ea65bbd4d566fa63476547 or the updated one in 810a33a9742688adf668367548b4adc114b84720 ?

dmsnell replied:
I like passing the arguments as actual parameters and not wrapped in an object unless they are normally optional parameters, so something like `( dispatch, action, getState )` over `( { dispatch, action, getState } )`

dmsnell commented on client/state/notices/middleware.js:
```diff
@@ -0,0 +1,62 @@
+		if ( text ) {
+			dispatch( successNotice( text ) );
+		}
+	}
```
these aren't really single-fine functions here for the most part, and they have their own independent behaviors. couldn't we define these functions independently and just map them into the object? I realize this would involve some duplicate code, but I find it easier to scan through and test

```js
export const showMessage = message => ( action, dispatch ) => dispatch( successNotice( message ) );
export const postWasSaved = ( action, dispatch ) => { let text; … };

const observers = [
	…,
	[ POST_DELETE_SUCCESS ]: showMessage( translate( 'Post successfully deleted' ) ),
	[ POST_SAVE_SUCCESS ]: postWasSaved,
	…
];
```

Now to test these things we don't have to import a big object of nothing but functions, rather we can just import those functions directly. Why would a _unit test_ even need to know that the handling code is inside some bigger data structure?

aduth replied:
>but I find it easier to scan through and test

I'd agree it'd make the list more scannable, but when I see code like how you'd described, I conversely find that I have trouble following the logic because my eyes are forced to bounce between the function declaration and its usage in the object.

>Why would a unit test even need to know that the handling code is inside some bigger data structure?

This is a good point, but what would be the difference between what you suggest and testing on `OBSERVERS[ POST_SAVE_SUCCESS ]`?

dmsnell replied:
> trouble following logic

It feels like we are just wired differently here, which is fine

> bounce between the function declaration and usage

Not sure what you are implying here. Isn't this one of the very reasons we use functions instead of writing one big flat source file? Maybe it's more about the extent to which we do this. I was struggling in this very same spot on my PR: where do I define inline and where do I separate. I feel like separating will be the need as these things get bigger (_ala_ the comments about file size) and it will allow us to use some reusable handling code (such as `showMessage`). While I don't care too much about small trees like this, I wonder about setting the precedent for others trying to learn what to do.

> difference between...

In the one case, we are testing a specific function for its behavior and that's all we need to care about. With the object wrapping around it, we're testing both the functionality _and_ the calling convention. We have to know about that convention and we have to make sure that we keep our tests in sync with any changes that come from changing the wrapper.

You could argue that this isn't a big deal, and I would probably concede that point. Nonetheless, I still find it useful to work towards a testing suite that doesn't tangle itself like that.

aduth replied:
>Isn't this one of the very reasons we use functions [...] Maybe it's more about the extent to which we do this.

Yes to both of these. It's a balance. I don't find your suggestion here to be particularly extreme, but I do think some of the [earlier discussion](https://github.com/Automattic/wp-calypso/pull/6645#discussion_r70321190) was starting to head in the direction of jumping the gun on abstractions (decorators, `createActionHandler`). My thought with keeping it as a single object was to keep it simple to start, since I'm honestly unsure how this will evolve (https://github.com/Automattic/wp-calypso/pull/6645#issuecomment-232050396). This may or may not scale very well, but there's not been an obvious alternative presented that would be more readable/maintainable.

>With the object wrapping around it, we're testing both the functionality and the calling convention.

Isn't this true of any variable? A variable's name is its calling convention, so I still don't see why we're any more tied to `HANDLERS[ POST_SAVE_SUCCESS ]` than we'd be to `handlePostSaveSuccess`.

dmsnell replied:
> jumping the gun on abstractions (decorators, createActionHandler)

I'm less a fan of those because I don't think they are the right abstractions than I am about jumping the gun, but yeah - basically in agreement and fine with the idea of starting small

dmsnell replied:
> Isn't this true of any variable? A variable's name is its calling convention

not really. functions defined using `function` and the fat-arrow syntax have a specific way of being invoked: add some parenthesis or use `call()` or `apply()`. What if we call `HANDLERS[ POST_HAS_NO_MORE_CARES_TO_GIVE ]`? What is the expected return of that function call? Well, it's nonsense because that's not a function. The key doesn't exist and thus no handler exists _at that key_, and even more, it's not a function because we didn't happen to set the value at that key _as a function_. The point is that when we have a function by itself, we've got it in full - no assumptions need to be made, no questions remain - it's ours. When we have an object, or a dispatching function of some kind, we have to trust that it's going to give us what we want and that there's no more behavior wrapping around the function call.

Again, this argument is more important not because of this particular bit of code, but in general because we don't have precedent and are trying to figure out what good style looks like in this situation.

dmsnell commented on client/state/notices/middleware.js:
```diff
@@ -0,0 +1,62 @@
+	POST_SAVE_SUCCESS
+} from 'state/action-types';
+
+export const OBSERVERS = {
```
WHY ARE YOU SHOUTING?

dmsnell replied:
also I'm kind of curious on your name `observer`. when I created something like this, my brainstorming led me to `handler` instead. the thinking was that since these functions get called by the middleware on certain events, they were thus event handlers.

I wouldn't normally care about the naming here, but _observer_ seems to carry with it some implications about _observables_ in general, as in I expect these to be _watching_ for something when in fact they don't do anything actively - all passively being called by something else. they're not even inserting a hook somewhere, just living in a data structure.

aduth replied:
>WHY ARE YOU SHOUTING?

Fun fact, the name for this style is SCREAMING_SNAKE_CASE.

More to the point, this is a [documented style for constant declarations](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#naming-conventions).

>also I'm kind of curious on your name `observer`.

I'm not opposed to `handler`. I think it popped into my head at least once in the course of implementing this. Given the concerns on interpretation, I'll just rename it.

dmsnell replied:
> documented style

It feels like that section could use some updating. This feels qualitatively different to me than something like is in the example:

```
const MAX_ITEMS = 23;
```

It feels incidentally constant as a side-effect of how we have chosen to organize the code. In fact, if we had used a switch structure instead this would be silly here. On the other hand, a literal constant is basically always a constant no matter where we use it.

Do we even use this outside of this file anywhere besides in the tests?

aduth replied:
>This feels qualitatively different to me [...]

Yeah, on considering this some more, I'm not sure the all-caps is appropriate here. To make matters more complicated, we (and be "we", I mostly mean "I" 😄 ) have used `PascalCase` in some instances of objects consisting only of constants ([example](https://github.com/Automattic/wp-calypso/blob/4d92cde41ca1ecc7bcf4e7809b7562a87c785b96/client/lib/media/constants.js#L12-L20)), which is already at odds with how I've implemented it here. I can't recall where specifically, but this came up as a recent discussion with @nylen on whether this was a good approach (and should be documented).

Do you think this is a distinction worth making (`Handlers`) or should we keep to our current guidelines and name as lowercase (`handlers`)?

dmsnell replied:
I prefer `Camel_case` (hehe) unless it's something of a universal constant, like `PI` or `PLANCK_MASS` or `POST_SAVE` or `TITLE_LIMIT`

nylen replied:
[Previous discussion](https://github.com/Automattic/wp-calypso/pull/6013#discussion_r68981097).  After thinking about it some more I like `PascalCase` the best for situations like this, but we should document that in a separate PR.  What are the applicable situations?  Collections of objects/constants/functions?  The pattern is almost like a class with static-only methods in languages that support such.

dmsnell replied:
Why `PascalCase`? What's wrong with just having it as a variable? For cases like this I don't consider them any different because I could just as easily see them as being defined inside of the function body as any other given variable. To change it here would feel like doing this inside of React `render()` bodies…

```js
render() {
	const POSTS = this.props.posts.filter( isPublic );
}
```

Do we really need to add another naming convention? I think it's bad enough that we have our PHP styles permeating Calypso because things aren't ever the way I expect them: it's JavaScript, so why isn't it `camelCase`?

Put another way, do we treat these variables in a significantly different enough way that it's justifiable to name them differently? How do we treat them different? What is their classification in contrast to normal variables or variables defined with `const`?

nylen replied:
> do we treat these variables in a significantly different enough way that it's justifiable to name them differently?

Maybe.. maybe not.  I'm glad we are discussing it :)

> What is their classification in contrast to normal variables or variables defined with `const`?

The best I could come up with is something analogous to "a class with static-only methods".  In the other PR we ended up going with a regular `camelCase` variable as you are advocating.

aduth replied:
A better example of where this is more commonly used is `ActionTypes`:

https://github.com/Automattic/wp-calypso/search?utf8=%E2%9C%93&q=ActionTypes

>Do we really need to add another naming convention? [...] do we treat these variables in a significantly different enough way that it's justifiable to name them differently?

I agree with the worry of adding a new convention. I think there's a small argument to be made that it helps indicate that an object contains only constants. I'm leaning more towards keeping it simple and not advocating this convention, and in any case it's probably best kept for a discussion separate from these changes. I'll plan to update this to `handlers`.

dmsnell replied:
> a class with static-only methods

Aha, @nylen - you have touched on another pet-peeve of mine :wink: Our PHP code is littered with static-only classes. That is, state-encapsulating containers which contain no state. I'm pretty sure that the only reason we have those in PHP all over the place is the general lack of namespacing for so long. At least here in JS we don't have that problem.

Point: If a class only has static methods, then we should probably eliminate the class and make those methods all functions instead.

aduth commented:
@dmsnell : Thanks for the detailed feedback. I've incorporated a number of your suggestions in 810a33a9742688adf668367548b4adc114b84720 . I've still not moved individual handlers to be separate from the object itself because I couldn't think of a nice pattern that satisfies the [worry I mentioned earlier](https://github.com/Automattic/wp-calypso/pull/6645#discussion_r70428913). At this point, it might be a bit premature anyways, and I'd be curious to see what trends arise for notice handler logic. If it's mostly like the `dispatchSuccess` case and few with extensive logic, it'd probably be reasonable to place them separate.

dmsnell commented:
there is nobody I am more intimidated by to leave a comment in a review than you @aduth - your level of quality and productivity is an example to us all. Thanks for interacting with me thus through these discussions. Please know hat even when I write about things for which I hold a strong opinion, I am also simply trying to challenge assumptions about the code and explore if there might be other ways to accomplish them - noting is intended personally. Further, I jumped in here because of how closely it relates to the work that @artpi and myself are doing. I'm excited to see a new specialized middleware come about and to see another developer working on his big problem.

aduth commented:
>there is nobody I am more intimidated by to leave a comment in a review than you @aduth

Ahh! I hate that you feel this way.

I'll never claim to always have it right, and as often as I disagree, these discussions are always formative for me. Code style is an fascinating topic and I'm very interested in understanding the benefits seen by other perspectives on the matter (ask @mcsf, we've discussed this at length recently 😄 ).

So please do continue to challenge me. As evidenced by how this has evolved from #6306, it's indeed a unique problem with a number of implementation approaches worth exploring.

aduth commented:
I've pushed another set of revisions in ca092b5 which hopefully addresses the remaining feedback. Specifically:

- Rename `HANDLERS` to `handlers`
- Move handler logic to separate functions
  - Mapping hopefully easier to scan with one-line-per-action-type
  - Testing is performed on the exported handler logic directly
- Returned to passing arguments directly, rather than as object
  - Ordered by most relevant to a handler
- A few comment blocks to help break up the file into sensible chunks
  - Maybe later we pull some of these out into their own files

blowery commented:
Ah! So much has happened. To respond to one thing I mentioned in a code diff with @dmsnell ...

>  I do not think that the reader should care or know anything about notices. why should it? the reader knows that it needs to tell Calypso to delete a post or something like that - end of story.

My concern with having all of the notice responders in one file has less to do with who-knows-what and more to do with bundle size. I agree the "reader" (ie, stuff that lives in client/reader) shouldn't know a darn thing about notices. I think the action creators _may_ know about them, but ideally they don't. So then who does know about them, and how do we decide which bundle they should get pulled into?

With the current layout, these will likely end up in the commons. That's probably ok, but as I mentioned, if one loads the reader and nothing else, code that handles actions that never fire (and _can_ never fire) also loads. This is true here, and for our reducer tree. Both are pulling in everything that could possibly happen from the start. I think this is unnecessary bloat and we're going to want / need a system to patch things into the reducer tree dynamically down the road. Our bundles are already pretty darn large. Making large parts of the app global isn't going to help that concern.

dmsnell commented:
> more to do with bundle size

this is a concern I haven't thought about @blowery. I suppose it's moot for notices since they are more like global actions anyway, but I see your point on others.

since these would all come in through middleware, it seems reasonable that we could possible bundle up the middleware into individual chunks or groups. we can dynamically load in middleware so maybe when we open *My Sites* (or when anticipating such a navigation) we could start pulling in other chunks.

there's no reason we couldn't split this middleware into smaller chunks than _WP API_. We could have _postMiddleware_, _siteMiddleware_, etc… though it would obviously take more thought than this to choose how to split it up.

dmsnell commented:
@aduth this is great stuff. I feel like those of us working with middleware approaches are searching for solid footing in an unstable place, but I guarantee you this is moving in the right direction and far better than what it is replacing. Thanks for the work, thanks for the diligence!

![](http://stream1.gifsoup.com/view7/2842621/make-it-so-o.gif)

blowery commented:
> since these would all come in through middleware, it seems reasonable that we could possible bundle up the middleware into individual chunks or groups. we can dynamically load in middleware so maybe when we open My Sites (or when anticipating such a navigation) we could start pulling in other chunks.

My thought for this particular thing was that the middleware would always be present, but code loaded via chunks could register handlers when they load.

The downside there is that if code that registers a handler gets duped into more than one chunk, we'd have a problem with the "same" code registering multiple times, resulting in multiple notices, maybe.

dmsnell commented:
> but code loaded via chunks could register handlers when they load

it would be worth getting some numbers on how much code the reducers currently consume and how much they would grow.

I'm assuming that right now we have lots of duplicated code lying around in _redux-thunks_ because of this lack of centrality, and I also assume that the reducers may not grow too big.

I will try and fiddle around with some numbers if I can find them.
